docs/gadget-devel/gadget-wasm-api-raw.md

```diff
@@ -280,3 +280,58 @@ Parameters:
 
 Return value:
 - None
+
+### Parameters
+
+#### `newParam(key string, description string, defaultValue string, typeHint string, title string, alias string, isMandatory u32) u32`
```

Comment:
Do we need `newParam`?

Wouldn't it be enough to define a parameter in gadget.yaml and add `setParam` in the wasm api?

Reply:
Some comments on this:

- Currently it's not that easy to do as the params on the metadata file are only handled by the ebpf operator [here](https://github.com/inspektor-gadget/inspektor-gadget/blob/d28c76d0ef853f40fa5594b24e3febf044b757d6/pkg/operators/ebpf/params.go#L105), it means those are only visible to the ebpf operator. 
- Even if we could define the parameters on the metadata file, I still think it's valuable to be able to create them from wasm:
  - The user doesn't have to bother defining the parameter in one place and reading it in another 
  - The parameter creation could use some complex logic that can't be specified in yaml

Reply:
> On the other hand, it means the --help documentation can only be printed by executing the wasm module, or at least the gadgetInit() function of the wasm module.

This is done similarly in other operators. For instance, the ebpf operator parses the eBPF spec to add parameters like [iface](https://github.com/inspektor-gadget/inspektor-gadget/blob/fbb6b2443e3abc2080d5c202cb4200ef932829c8/pkg/operators/ebpf/ebpf.go#L505-L510). IMHO it's ok to call these functions to generate the help, in the end that's the source of truth regarding the available flags according to the env where the gadget is running.

> Also, the template gadgets/README-flags.template and the make -C gadgets readme would need to run the wasm module too. I would prefer to documentation directly available in gadget.yaml, so it can easily be read by third-party tools.

I have some concerns with auto generating the list of flags from the metadata file:
- The documentation don't include any flag provided by the different operators (--filter, --containername, --iface, etc). 
- What about flags that could be added dynamically based on a specific condition? Or that it's default value can be changed based on something?

In general I think third party tools (like UIs) will have to rely more in the `GetGagetInfo()` to be able to provide all parameters (flags) that are supported by a gadget in a specific environment. Perhaps we can use this also to generate the flags instead of only relying only on the metadata?

> With
https://github.com/inspektor-gadget/inspektor-gadget/pull/3475, both api.SetAnnotation and api.NewParam are called from the same gadgetInit() function.
With
https://github.com/inspektor-gadget/inspektor-gadget/issues/3131, there will be more functions doing low-level calls like reading kallsyms, using bpf() syscall. If there are done in the same gadgetInit() function, that means the documentation can only be generated on a Linux machine.

I actually have to introduce two new PreStart() and PostStop() functions in #3463. We can (and should) document that the Init() one is called in preparation (also when running `--help`) stating that the gadget developer has to be careful not to perform heavy operations there. 

> If you still want to have it defined in the wasm module, I think it should be in a dedicated function gadgetAddParams, so that --help does not need to call gadgetInit() doing low-level calls.

Please see the comment above above PreStart / PostStop. Those heavy operations could be done on PreStart().

Reply:
For parameters defined in the eBPF code, we have `--update-metadata` to read them from the eBPF module and copy them in gadget.yaml. Do you think we can have something similar for the parameters defined in the WASM module?

I realise it does not work for parameters dynamically generated from the current environment (e.g. kallsyms, list of LSM hooks, etc.). But in my opinion, that's something to avoid. In a Kubernetes cluster with several nodes potentially using different kernel versions, it could potentially generate different parameters. When using `kubectl-gadget run` to get the help, it could give a different answer depending on the node that executed the WASM module.

> What about flags that could be added dynamically based on a specific condition? Or that it's default value can be changed based on something?

In #3463 (trace_fsslower), the parameter "filesystem" is not dynamic: it is just a string with the default value "ext4" (a string). So I don't know if we have a real use case where we need dynamic parameters that are defined differently depending on the environment.

It could also have the value "all" where it attaches the kprobes to all supported filesystems on the current node. So an heterogeneous cluster with some nodes with xfs support and some without, it would be possible to use `--filesystem=all` to attach it to whichever filesystems are supported on each node. No need for dynamic parameters for this use case.

Reply:
I removed the newParam function and switched to use the metadata. We can implement it back if needed later on.